### PR TITLE
RDP: Fix ACK unconditionally accepted via RST + ACK

### PR DESCRIPTION
**Pull request by moonlight83340**
If a RST + ACK segment arrives in the OPEN state, the ACK was accepted unconditionally. 
Add a sequence number check to verify if the RST-ACK is out of sequence.

See issue https://github.com/libcsp/libcsp/issues/369

### CODE REVIEW

File: src/csp_rdp.c [hunk 1]
```diff
@@ -523,8 +523,13 @@ bool csp_rdp_new_packet(csp_conn_t * conn, csp_packet_t * packet) {
 	if (rx_header->flags & RDP_RST) {
 
 		if (rx_header->flags & RDP_ACK) {
-			/* Store current ack'ed sequence number */
-			conn->rdp.snd_una = rx_header->ack_nr + 1;
+			if (rx_header->seq_nr == (conn->rdp.rcv_cur + 1)){
+				/* Store current acked sequence number */
+				conn->rdp.snd_una = rx_header->ack_nr + 1;
+			}else{
```
Comment (yashi) [inline src/csp_rdp.c]: fix white space

File: src/csp_rdp.c [hunk 2]
```diff
@@ -523,8 +523,13 @@ bool csp_rdp_new_packet(csp_conn_t * conn, csp_packet_t * packet) {
...
+				csp_rdp_protocol("RDP %p: RST ACK out of sequence, keep connection open\n", conn);
```
Comment (yashi) [inline src/csp_rdp.c]: print received and expected sequence numbers

**Comment by moonlight83340** [comment]
Rebase.

File: src/csp_rdp.c [hunk 3]
```diff
@@ -529,8 +529,13 @@ bool csp_rdp_new_packet(csp_conn_t * conn, csp_packet_t * packet) {
 	if (rx_header->flags & RDP_RST) {
 
 		if (rx_header->flags & RDP_ACK) {
-			/* Store current ack'ed sequence number */
-			conn->rdp.snd_una = rx_header->ack_nr + 1;
+			if (rx_header->seq_nr == (conn->rdp.rcv_cur + 1)) {
+				/* Store current acked sequence number */
+				conn->rdp.snd_una = rx_header->ack_nr + 1;
+			} else {
+				csp_rdp_protocol("RDP %p: RST ACK out of sequence, seq_nr : %d, expected seq_nr: %d,keep connection open\n", conn, rx_header->seq_nr, (conn->rdp.rcv_cur + 1));
```
Comment (yashi) [inline src/csp_rdp.c]: the line is too long.  Move arguments to the next line.

**Review by yashi — CHANGES_REQUESTED** [review]

**Comment by yashi** [comment]
Please elaborate a bit more.

This is the original commit log:

> If a RST + ACK segment arrives in the OPEN state,
> the ACK was accepted unconditionally.
> Add a sequence number check to verify if the RST-ACK is out of sequence.

It doesn't say
- why it accept unconditionally
- why it's a bad thing

Also, 
- is "a RST + ACK segment" the same as "the ACK"? (what the old code does is `/* Store current ack'ed sequence number */`)
- With the commit, you only accept `rx_header->seq_nr == (conn->rdp.rcv_cur + 1)`, but what if order of the arriving packet is not sequential?

File: src/csp_rdp.c [hunk 4]
```diff
@@ -529,8 +529,14 @@ bool csp_rdp_new_packet(csp_conn_t * conn, csp_packet_t * packet) {
 	if (rx_header->flags & RDP_RST) {
 
 		if (rx_header->flags & RDP_ACK) {
-			/* Store current ack'ed sequence number */
-			conn->rdp.snd_una = rx_header->ack_nr + 1;
+			if (rx_header->seq_nr == (conn->rdp.rcv_cur + 1)) {
```
Comment (yashi) [inline src/csp_rdp.c:532]: What if packets arrives non sequential order?

Comment (moonlight83340) [reply src/csp_rdp.c:532]: The RST will be ignored.

Comment (yashi) [reply src/csp_rdp.c:532]: That's not acceptable, then.

Comment (yashi) [reply src/csp_rdp.c:532]: Is it acceptable as RDP?

Comment (moonlight83340) [reply src/csp_rdp.c:532]: https://www.rfc-editor.org/rfc/rfc908
https://datatracker.ietf.org/doc/html/rfc1151

I haven't found in the RFC if it's acceptable or not to do so.
I have the same impression as the author, I think we should not accept an RST ACK packet that arrives in the wrong order. (a sequence number different from that expected)

Comment (yashi) [reply src/csp_rdp.c:532]: care to elaborate a bit more?

**Comment by johandc** [comment]
Hello, there is already a check on RST out of sequence, that sends to discard_open.

```
	/* If a RESET was received. */
	if (rx_header->flags & RDP_RST) {

		if (rx_header->flags & RDP_ACK) {
			/* Store current ack'ed sequence number */
			conn->rdp.snd_una = rx_header->ack_nr + 1;
		}

		if (conn->rdp.state == RDP_CLOSED) {
			csp_rdp_protocol("RDP %p: RST received in CLOSED - ignored\n", conn);
			close_connection = (conn->dest_socket != NULL);
			goto discard_open;
		}

		if (conn->rdp.state == RDP_CLOSE_WAIT) {
			csp_rdp_protocol("RDP %p: RST received in CLOSE_WAIT, ack: %d - closing\n", conn, (rx_header->flags & RDP_ACK));
			if ((rx_header->flags & RDP_ACK) && CSP_USE_RDP_FAST_CLOSE) {
				// skip timeout - the other end has acknowledged the RST
				closed_by |= CSP_RDP_CLOSED_BY_TIMEOUT;
			}
			goto discard_close;
		}

		if (rx_header->seq_nr == (conn->rdp.rcv_cur + 1)) {
			csp_rdp_protocol("RDP %p: Received RST in sequence, no more data incoming, reply with RST\n", conn);
			conn->rdp.state = RDP_CLOSE_WAIT;
			conn->timestamp = csp_get_ms();
			csp_rdp_send_cmp(conn, NULL, RDP_ACK | RDP_RST, conn->rdp.snd_nxt, conn->rdp.rcv_cur);
			if (CSP_USE_RDP_FAST_CLOSE) {
				closed_by |= CSP_RDP_CLOSED_BY_TIMEOUT;
			}
			goto discard_close;
		}

		csp_rdp_protocol("RDP %p: RST out of sequence, keep connection open\n", conn);
		goto discard_open;
	}
```

The ACK_NR field should always be unconditionally accepted. Because its conveying valid information on the stream in the other direction.

**Pull request closed without merging**